frame/support/procedural/src/storage/storage_struct.rs

```diff
@@ -207,9 +207,107 @@ pub fn decl_and_impl(scrate: &TokenStream, def: &DeclStorageDefExt) -> TokenStre
 			}
 		};
 
+		let max_values = if let Some(max_values) = &line.max_values {
+			quote::quote!({
+				let max_values: u32 = (|| #max_values)();
+				Some(max_values)
+			})
+		} else {
+			quote::quote!(None)
+		};
+
+		let storage_info_impl = if def.generate_storage_info {
+			match &line.storage_type {
+				StorageLineTypeDef::Simple(_) => {
+					quote!(
+						impl<#impl_trait> #scrate::traits::StorageInfoTrait for #storage_struct
+						#optional_storage_where_clause
+						{
+							fn storage_info() -> #scrate::traits::StorageInfo {
+								use #scrate::sp_runtime::SaturatedConversion;
+
+								let max_size = <
+									#value_type as #scrate::traits::MaxEncodedLen
+								>::max_encoded_len()
+									.saturated_into();
+								#scrate::traits::StorageInfo {
+									prefix: <
+										#storage_struct as #scrate::#storage_generator_trait
+									>::storage_value_final_key(),
+									max_values: Some(1),
+									max_size: Some(max_size),
+								}
+							}
+						}
+					)
+				},
+				StorageLineTypeDef::Map(map) => {
+					let key = &map.key;
+					quote!(
+						impl<#impl_trait> #scrate::traits::StorageInfoTrait for #storage_struct
+						#optional_storage_where_clause
+						{
+							fn storage_info() -> #scrate::traits::StorageInfo {
+								use #scrate::sp_runtime::SaturatedConversion;
+								let max_size = <
+									#value_type as #scrate::traits::MaxEncodedLen
+								>::max_encoded_len()
+									.saturating_add(
+										<#key as #scrate::traits::MaxEncodedLen>::max_encoded_len()
+									)
+									.saturated_into();
+								#scrate::traits::StorageInfo {
+									prefix: <
+										#storage_struct
+										as #scrate::storage::StoragePrefixedMap<#value_type>
+									>::final_prefix(),
+									max_values: #max_values,
+									max_size: Some(max_size),
```

Comment:
Not sure we have the right granularity here.

The impact on the child trie for a map will be some multiple of the value size * max_values

Since `max_size` encodes both the key and value, i dont think we can do the right math here to get the right PoV impact

@cheme

Reply:
"child trie" ? I guess you mean the branched portion of the trie containing the Map.

The fact that `max_size` encodes both key and value is correct for iterable maps I think (IIRC key is stored with value as trie node value), and not for non iterable one I think.

So from 'max_values' you can estimate the number of branches in PoV. Size of branch also could be estimate since the trie key are fix length hash.
Depending on the number of value accessed  (and kind of access), you will can estimate the number of shared branch too.

Reply:
> and not for non iterable one I think.

Ah you are right it is only the hasher which concatenate the key.

Reply:
Should we include the hash part when computing the max_size for a storage key/value ?

like if a storagemap with hashertwox64concat has only one value, then the node contains the full key and the value, so the size include the hash part. (so 8 bytes + key::max_encoded_len + value::max_encoded_len).

What do you think ?
I am currently implementing this.

Reply:
I would share the 8bytes amongst branches size evaluation and then do (key::max_encoded_len + value::max_encoded_len) as you suggest (doing as if all the encoded key is in the leaf node (worst case scenario)).

Reply:
Looking again at what is  hasertwox64, (twox64(key), key) would make unhashed part highly improbably to be in a branch so it is no really a worst case scenario.
Putting key with value seems correct. (and twox part shared amongst branches).

Reply:
> highly improbably

Yeah I agree, I'm ok to remove the hash part, if it is considered important we can improve on it later